tests/sundials/n_vector.cc

```diff
@@ -0,0 +1,625 @@
+    LinearAlgebra::distributed::BlockVector<double> vector(partitioning,
+                                                           MPI_COMM_WORLD);
+    vector = value;
+    return vector;
+  }
+
+
+  template <>
+  TrilinosWrappers::MPI::Vector
+  create_test_vector(double value)
+  {
+    IndexSet local_dofs = create_parallel_index_set();
+
+    TrilinosWrappers::MPI::Vector vector(local_dofs, MPI_COMM_WORLD);
+    vector = value;
+    return vector;
+  }
+
+  template <typename VectorType>
+  bool
+  vector_equal(const VectorType &a, const VectorType &b)
+  {
+    auto elements_a = a.locally_owned_elements();
+    auto elements_b = b.locally_owned_elements();
+    if (elements_a.size() != elements_b.size())
+      return false;
+
+    const auto is_equal = [&](const IndexSet::size_type idxa,
+                              const IndexSet::size_type idxb) {
+      return std::fabs(a[idxa] - b[idxb]) < 1e-15;
+    };
+
+    return std::equal(elements_a.begin(),
+                      elements_a.end(),
+                      elements_b.begin(),
+                      is_equal);
+  }
+} // namespace
+
+
+template <typename VectorType>
+void
+test_nvector_view_unwrap()
+{
+  auto                    vector       = create_test_vector<VectorType>();
+  const auto              const_vector = create_test_vector<VectorType>();
+  NVectorView<VectorType> n_vector(vector);
+  NVectorView<VectorType> const_n_vector(const_vector);
+
+  Assert(n_vector != nullptr, NVectorTestError());
+  Assert((n_vector)->content != nullptr, NVectorTestError());
+
+  Assert(const_n_vector != nullptr, NVectorTestError());
+  Assert((const_n_vector)->content != nullptr, NVectorTestError());
+
+  auto *vector_unwrapped = unwrap_nvector<VectorType>(n_vector);
+  // here we  check for pointer equality
+  Assert(vector_unwrapped == &vector, NVectorTestError());
+
+  // unwrap non-const as const
+  const auto *const_vector_unwrapped =
+    unwrap_nvector_const<VectorType>(n_vector);
+  Assert(const_vector_unwrapped == &vector, NVectorTestError());
+
+  // unwrap const vector as const
+  const_vector_unwrapped = unwrap_nvector_const<VectorType>(const_n_vector);
+  Assert(const_vector_unwrapped == &const_vector, NVectorTestError());
+
+  // unwrap const vector as non-const throws
+  try
+    {
+      unwrap_nvector<VectorType>(const_n_vector);
+      Assert(false, NVectorTestError());
+    }
+  catch (ExcMessage e)
+    {
+      const std::string msg(e.what());
+      Assert(msg.find(
+               "Tried to access a constant vector content as non-const.") !=
+               std::string::npos,
+             NVectorTestError());
+    }
+
+  deallog << "test_nvector_view_unwrap OK" << std::endl;
+}
+
+
+
+template <typename VectorType>
+void
+test_get_vector_id()
+{
+  auto                    vector = create_test_vector<VectorType>();
+  NVectorView<VectorType> n_vector(vector);
+  auto                    id = N_VGetVectorID(n_vector);
```

Comment:
These lines demonstrate the usage of `NVectorView`. After creating the view, it can be passed to a SUNDIALS function where it implicitly converts to `N_Vector`. The view is cleaned up when going out of scope.